index.js

```diff
@@ -106,6 +118,11 @@
     return f(this.value);
   };
 
+  //  Just#concat :: m a -> m a
+  Just.prototype.concat = function(maybe) {
+    return maybe instanceof Just ? Just(this.value.concat(maybe.value)) : this;
+  };
+
```

Comment:
>>  I want my Maybe to be as bullet-proof as possible.

> That's an admirable goal. I think the only way to really make that happen is to stop writing vanilla js, or use a different abstraction. `Maybe` is too low an abstraction to protect you from all the possible ways things can go wrong.

I wasn't expecting it to solve all my problems, mostly to reduce `null`-checking.  Javascript code that's full of `null`-checks can be more nicely written with the help of some judiciously-placed `Maybe`s.

I did understand the idea of writing an abstract implementation that would work for any conforming types.  But my reference for the ADTs in Javascript has long been FantasyLand, which makes no mention of any interaction between `Semigroup`/`Monoid` laws and laws of the types in `Monad`'s genealogy.  While the laws you discuss here make a great deal of sense, this is the first I've heard of them.

And I still have significant reservations because although the `filter` function does not actually use `concat`, the `concat` proposed for `Maybe` seems designed for `Monoid a => Monoid (Maybe a)` (likely just `Semigroup`, but `Monoid` seems more likely useful.)  However I see nothing in the Sanctuary implementation that says that these `Maybe`s are more strongly typed than `Maybe *`.  Declaring a `Semigroup`/`Monoid` on `Maybe *` doesn't make sense to me.  And yet the notion of `filter` **does** make sense for `Maybe *`, with a fairly obvious implementation:

```javascript
Nothing.prototype.filter = function(predicate) {return this;}
Just.prototype.filter = function(predicate) {
    return predicate(this.value) ? this : new Nothing();
}
```

I do understand that this loses out by not taking advantage of generic laws.  But it seems one could only do so by pushing more of an external type system into Javascript.

Am I making sense here?  Or is my lack of experience with languages like Haskell still making me miss something fundamental?  I do have experience with Monoids, Semigroups, etc., but from a mathematical, not a programming point of view.  I did take one course in Category Theory in grad school; it was my least favourite course and helped steer me to more discrete fields; my work was in combinatorics and graph theory.  I really remember nothing of Monads or some of these other types.  And I have less feel for them than I do for the sorts of types that make up an abstract algebra course.